***UNIVERSIDAD CENTRAL DEL ECUADOR***

***FACULTAD DE CIENCIAS ECONÓMICAS***


---



***Curso de Python (Investiga y aprende)***


---

**Estudiante:**

*Bauz Mamallacta Nicole Jenny*


---



**Clase 1 / Ejercicios**


---


**1.   Suponga que tiene una lista [1, 2, 3, 2, 1] y desea crear una nueva lista sin números dobles, después de resolver este problema, intente reescribirla como una sola línea.**

*Un one liner en programación es un programa existente de una sola línea. Si no logras escribir una sola línea, no te preocupes. el codigo one liner es a menudo difícile de leer y, como hemos aprendido esta semana: los lenguajes de programación están destinados a ser legibles para las personas. Así que no te preocupes si estás usando demasiado código para resolver un problema, debes preguntarte: ¿yo o alguien más entenderá esto dentro de 3 años?*

In [4]:
# Primer Paso: Creación de la lista
list=[1,2,3,2,1]
print(list)

[1, 2, 3, 2, 1]


In [5]:
# Segundo Paso: Solución del problema
rep=set(list) # Set: Elimina los valores repetidos
ord=sorted(rep) # Sorted: Ordena la lista de forma ascendente
print(ord) # Print: Muestra la solución del problema (Valores únicos)

[1, 2, 3]


**2.   Slicing puede tener 3 parámetros, inicio, final y paso. Sin embargo, a menudo escribirá o leerá operaciones de corte con menos de 3 parámetros. En esta pregunta, debe volver a escribir todos los ejemplos en operaciones de corte con 3 parámetros. Después de haberlas reescrito, verifique si sus respuestas dan los mismos resultados**



In [8]:
string_1 = 'abcde'
print(string_1[1:3])
print(string_1[-3:-2])

print(string_1[:3])
print(string_1[3:])
print(string_1[:])

print(string_1[-2:])
print(string_1[:-2])

print(string_1[::])
print(string_1[::-1])

print(string_1[:3:1])
print(string_1[:3:-1])

print(string_1[1::])
print(string_1[:1:-1])

bc
c
abc
de
abcde
de
abc
abcde
edcba
abc
e
bcde
edc


In [7]:
# Operación 1
# Usando Slicing básico: inicio y fin
string_1 = 'NICOLE'
print(string_1[1:3])
print(string_1[-3:-2])

print(string_1[:3])
print(string_1[3:])
print(string_1[:])

print(string_1[-2:])
print(string_1[:-2])

print(string_1[::])
print(string_1[::-1])

print(string_1[:3:1])
print(string_1[:3:-1])

print(string_1[1::])
print(string_1[:1:-1])

IC
O
NIC
OLE
NICOLE
LE
NICO
NICOLE
ELOCIN
NIC
EL
ICOLE
ELOC


In [10]:
# Operación 2
# Usando Slicing básico: inicio, fin y paso
string_1 = 'NICOLE'
print(string_1[1:3:])
print(string_1[-3:-2:])

print(string_1[:3:])
print(string_1[3::])
print(string_1[::])

print(string_1[-2::])
print(string_1[:-2:])

print(string_1[::])
print(string_1[::-1])

print(string_1[:3:])
print(string_1[:3:-1])

print(string_1[1::])
print(string_1[:1:-1])

IC
O
NIC
OLE
NICOLE
LE
NICO
NICOLE
ELOCIN
NIC
EL
ICOLE
ELOC


**Respuesta:** *Efectivamente, las respuestas son las mismas utilizando las diferentes operaciones.*

***3.   Hemos construido un diccionario que se puede usar para encontrar el resultado si dos booleanos están conectados con el operador y
Reescribe este diccionario para encontrar el resultado si dos booleanos están conectados con el operador.***

Python no tiene un operador XOR, busque en Internet lo que se supone que debe hacer el operador XOR y cree un diccionario que proporcione el resultado de una operación XOR


In [12]:
# Primer Operador
age = 40
blood_type = 'A'
my_and = {(True, True): True, (True, False): False, (False, True): False, (False, False): False }
print(my_and[age < 70, blood_type == 'B'])

False


In [13]:
# Segundo Operador
age = 40
blood_type = 'A'
my_and = {(True, True): True, (True, False): True, (False, True): True, (False, False): False }
print(my_and[age < 70,blood_type == 'B'])

True


**Respuesta:** *En el primer operador utilizamos AND y en el segundo utilizamos el OR, por esa razón el resultado de la ejecución de cada operador es diferente (Contraria a la opuesta).*

***4.   En 'Las listas pueden contener listas y así formar tablas de múltiples dimensiones' se les mostro una secuencia ganadora en tres en raya. Como es muy importante para el análisis de datos y la visualización de datos que se sienta cómodo trabajando con tablas poli-dimensionales, le pedimos que reescriba la secuencia con un segundo jugador más inteligente (o un jugador que aprendió el juego de memoria) donde el juego termina en un empate.***


In [ ]:
tic_tac_toe_board = [[' ', ' ', ' '], [' ', ' ', ' '], [' ', ' ', ' ']]
print(tic_tac_toe_board[0]) 
print(tic_tac_toe_board[1]) 
print(tic_tac_toe_board[2])
print() 
tic_tac_toe_board[1][1] = 'X'
print(tic_tac_toe_board[0]) 
print(tic_tac_toe_board[1]) 
print(tic_tac_toe_board[2])
print() 
tic_tac_toe_board[1][2] = 'O'
print(tic_tac_toe_board[0]) 
print(tic_tac_toe_board[1]) 
print(tic_tac_toe_board[2])
print()
tic_tac_toe_board[0][2] = 'X'
print(tic_tac_toe_board[0]) 
print(tic_tac_toe_board[1]) 
print(tic_tac_toe_board[2])
print()
tic_tac_toe_board[2][0] = 'O'
print(tic_tac_toe_board[0]) 
print(tic_tac_toe_board[1]) 
print(tic_tac_toe_board[2])
print()
tic_tac_toe_board[0][0] = 'X'
print(tic_tac_toe_board[0]) 
print(tic_tac_toe_board[1]) 
print(tic_tac_toe_board[2])
print()# And the next move X wins independent what player O plays

In [ ]:
# Ejercicio: Los jugadores queden empates
tic_tac_toe_board = [[' ', ' ', ' '], [' ', ' ', ' '], [' ', ' ', ' ']]
# Solución:
# Inicio del Juego
print(tic_tac_toe_board[0]) 
print(tic_tac_toe_board[1]) 
print(tic_tac_toe_board[2])
print() 

# Primer movimiento

tic_tac_toe_board[1][1] = 'X'
print(tic_tac_toe_board[0]) 
print(tic_tac_toe_board[1]) 
print(tic_tac_toe_board[2])
print() 

# Segundo movimiento.

tic_tac_toe_board[2][2] = 'O'
print(tic_tac_toe_board[0]) 
print(tic_tac_toe_board[1]) 
print(tic_tac_toe_board[2])
print()

# Tercer movimiento.

tic_tac_toe_board[0][0] = 'X'
print(tic_tac_toe_board[0]) 
print(tic_tac_toe_board[1]) 
print(tic_tac_toe_board[2])
print()

# Cuarto movimiento.

tic_tac_toe_board[1][0] = 'O'
print(tic_tac_toe_board[0]) 
print(tic_tac_toe_board[1]) 
print(tic_tac_toe_board[2])
print()

# Quinto movimiento.

tic_tac_toe_board[2][0] = 'X'
print(tic_tac_toe_board[0]) 
print(tic_tac_toe_board[1]) 
print(tic_tac_toe_board[2])
print()

# Sexto movimiento.

tic_tac_toe_board[2][1] = 'O'
print(tic_tac_toe_board[0]) 
print(tic_tac_toe_board[1]) 
print(tic_tac_toe_board[2])
print()

# Septimo movimiento.

tic_tac_toe_board[1][2] = 'X'
print(tic_tac_toe_board[0]) 
print(tic_tac_toe_board[1]) 
print(tic_tac_toe_board[2])
print()

# Octavo movimiento.

tic_tac_toe_board[0][2] = 'O'
print(tic_tac_toe_board[0]) 
print(tic_tac_toe_board[1]) 
print(tic_tac_toe_board[2])
print()

# Noveno movimiento.

tic_tac_toe_board[0][1] = 'X'
print(tic_tac_toe_board[0]) 
print(tic_tac_toe_board[1]) 
print(tic_tac_toe_board[2])
print()

***5. Ejecute el siguiente programa, vea si puede encontrar algo extraño y explique el resultado usando su conocimiento de recolección de basura.***

In [14]:
list_5 = [1, 2, 3]
print(f"The value of the list refered to by list_5 is {list_5} and its address is {id(list_5)}")
list_5 = [1 , 1, 3]
print(f"The value of the list refered to by list_5 is {list_5} and its address is {id(list_5)}")
list_5 = [1, 1, 4]
print(f"The value of the list refered to by list_5 is {list_5} and its address is {id(list_5)}")

The value of the list refered to by list_5 is [1, 2, 3] and its address is 140040404633872
The value of the list refered to by list_5 is [1, 1, 3] and its address is 140040465799200
The value of the list refered to by list_5 is [1, 1, 4] and its address is 140040404633872


**Respuesta:** *Se puede observar que la lista_5 = [1, 2, 3] tiene como id (140040404633872) y la list_5 = [1, 1, 4] tambien tiene el mismo id (140040404633872). Esto se resume en que los dos id, tanto de la lista_5 = [1, 2, 3], como de la list_5 = [1, 1, 4] son iguales.*


***6. Las cadenas son inmutables, pero ¿por qué funciona lo siguiente? ¿Claramente cambiamos el valor de la cadena llamada string_1? ¿Qué afirmaciones agregaría para aclarar lo que ha sucedido? usar id.***

In [15]:
string_1 = "text 1"
print(string_1)
string_1 = "text 2"
print(string_1)

text 1
text 2


In [19]:
string_1 = "text 1"
print(id(string_1))
string_1 = "text 2"
print(id(string_1))

140040403961200
140040404130032


**Respuesta:** *Esto sucede debido al intercambio de variables. Python no puede leer dos veces string_1, se estra sobreescibiendo la variable objeto. La cadena es inmutable*
